File: week6/Chapter_8_Refactoring/csharp/src/OopOrganization.Chapter8/Refactoring/LineItems.cs
namespace OopOrganization.Refactoring;

public interface ILineItem
{
    Money ExtendedPrice();
    bool IsShippable { get; }
}

public sealed record ProductLine(string Sku, int Qty, Money UnitPrice) : ILineItem
{
    public Money ExtendedPrice() => UnitPrice.Mul(Qty);
    public bool IsShippable => true;
}

public sealed record ServiceLine(string Name, decimal Hours, Money Rate) : ILineItem
{
    public Money ExtendedPrice() => Rate.Mul(Hours);
    public bool IsShippable => false;
}

File: week6/Chapter_8_Refactoring/javascript/src/afterRefactor/lineItems.js
export class LineItem {
  extendedPrice() { throw new Error("Not implemented"); }
  get isShippable() { throw new Error("Not implemented"); }
}

export class ProductLine extends LineItem {
  constructor(sku, qty, unitPrice) {
    super();
    this.sku = sku; this.qty = qty; this.unitPrice = unitPrice;
  }
  extendedPrice() { return this.unitPrice.mul(this.qty); }
  get isShippable() { return true; }
}

export class ServiceLine extends LineItem {
  constructor(name, hours, rate) {
    super();
    this.name = name; this.hours = hours; this.rate = rate;
  }
  extendedPrice() { return this.rate.mul(this.hours); }
  get isShippable() { return false; }
}
